File: health_monitor/examples/deadline_monitor.cpp
// Copyright (c) 2025 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache License Version 2.0 which is available at
// <https://www.apache.org/licenses/LICENSE-2.0>
//
// SPDX-License-Identifier: Apache-2.0
//

#include <cassert>
#include <chrono>
#include <cstdlib>
#include <iostream>
#include <mutex>
#include <thread>
#include <ffi/deadline_monitor.hpp>

std::mutex on_status_changed_lock;

int main()
{
    auto *builder = hm_dmb_new();

    hm_DeadlineMonitor *monitor;
    hm_dmb_build(&builder, &monitor);

    assert(builder == nullptr);
    assert(hm_dm_status(monitor) == hm_Status::Running);

    hm_dm_disable(monitor);
    assert(hm_dm_status(monitor) == hm_Status::Disabled);

    // Disable while disabled. Does nothing.
    hm_dm_disable(monitor);  // TODO: Should report error and the error should be handled.
    assert(hm_dm_status(monitor) == hm_Status::Disabled);

    hm_dm_enable(monitor);
    assert(hm_dm_status(monitor) == hm_Status::Running);

    // Enable while enabled. Does nothing.
    hm_dm_enable(monitor);  // TODO: Should report error and the error should be handled.
    assert(hm_dm_status(monitor) == hm_Status::Running);

    using namespace std::chrono_literals;

    std::thread t1(
        [&]()
        {
            hm_Deadline *deadline1;
            hm_dm_create_custom_deadline(monitor, 10, 1000, &deadline1);
            hm_Deadline *deadline2;
            hm_dm_create_custom_deadline(monitor, 50, 250, &deadline2);

            // Run task 1.
            hm_dl_start(deadline1);
            std::this_thread::sleep_for(250ms);
            hm_dl_stop(deadline1);

            assert(hm_dm_status(monitor) == hm_Status::Running);

            // Run task 2.
            hm_dl_start(deadline2);
            std::this_thread::sleep_for(100ms);
            hm_dl_stop(deadline2);

            assert(hm_dm_status(monitor) == hm_Status::Running);

            // Run task 1 again.
            hm_dl_start(deadline1);
            std::this_thread::sleep_for(250ms);
            hm_dl_stop(deadline1);

            assert(hm_dm_status(monitor) == hm_Status::Running);

            hm_dl_delete(&deadline1);
            assert(deadline1 == nullptr);
            hm_dl_delete(&deadline2);
            assert(deadline2 == nullptr);
        });

    t1.join();

    std::thread t2(
        [&]()
        {
            hm_Deadline *deadline;
            hm_dm_create_custom_deadline(monitor, 0, 100, &deadline);

            // This task is too long.
            hm_dl_start(deadline);
            std::this_thread::sleep_for(250ms);
            hm_dl_stop(deadline);

            assert(hm_dm_status(monitor) == hm_Status::Failed);

            hm_dl_delete(&deadline);
            assert(deadline == nullptr);
        });

    t2.join();

    assert(hm_dm_status(monitor) == hm_Status::Failed);

    hm_dm_delete(&monitor);
    assert(monitor == nullptr);

    return EXIT_SUCCESS;
}

File: health_monitor/examples/deadline_monitor.rs
// Copyright (c) 2025 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache License Version 2.0 which is available at
// <https://www.apache.org/licenses/LICENSE-2.0>
//
// SPDX-License-Identifier: Apache-2.0
//
use health_monitor::{common::*, deadline_monitor::*};
use std::{thread, time::Duration};

fn main() {
    let deadline_monitor_builder = DeadlineMonitorBuilder::new();
    let deadline_monitor = deadline_monitor_builder
        .build()
        .expect("Failed to build the monitor.");

    let deadline_monitor_clone_1 = deadline_monitor.clone();
    let t_1 = thread::spawn(move || {
        let mut deadline_1 = deadline_monitor_clone_1
            .create_custom_deadline(DurationRange::from_millis(10, 1000))
            .unwrap();
        let mut deadline_2 = deadline_monitor_clone_1
            .create_custom_deadline(DurationRange::from_millis(50, 250))
            .unwrap();

        // Run task 1.
        deadline_1.start().expect("Failed to start.");
        thread::sleep(Duration::from_millis(250));
        deadline_1.stop().expect("Failed to stop.");

        // Run task 2.
        deadline_2.start().expect("Failed to start.");
        thread::sleep(Duration::from_millis(100));
        deadline_2.stop().expect("Failed to stop.");

        // Run task 1 again.
        deadline_1.start().expect("Failed to start.");
        thread::sleep(Duration::from_millis(250));
        deadline_1.stop().expect("Failed to stop.");
    });

    let deadline_monitor_clone_2 = deadline_monitor.clone();
    let t_2 = thread::spawn(move || {
        let mut deadline = deadline_monitor_clone_2
            .create_custom_deadline(DurationRange::from_millis(10, 1000))
            .unwrap();

        deadline.start().expect("Failed to start.");
        thread::sleep(Duration::from_millis(250));
        deadline.stop().expect("Failed to stop.");
    });

    let _ = t_1.join();
    let _ = t_2.join();

    assert_eq!(deadline_monitor.status(), Status::Running);

    let deadline_monitor_clone_3 = deadline_monitor.clone();
    let t_3 = thread::spawn(move || {
        let mut deadline = deadline_monitor_clone_3
            .create_custom_deadline(DurationRange::from_millis(0, 100))
            .unwrap();

        // This task is too long.
        deadline.start().expect("Failed to start.");
        thread::sleep(Duration::from_millis(250));
        deadline.stop().expect_err("Stop should have failed.");
    });

    let _ = t_3.join();

    assert_eq!(deadline_monitor.status(), Status::Failed);
}
